### PR TITLE
Add Mosquitto tuning

### PR DESCRIPTION
**Pull request by Imaanpreet**
* [x] I am familiar with the [contributing](https://github.com/theforeman/foreman-documentation/blob/master/CONTRIBUTING.md) guidelines.

Please cherry-pick my commits into:

* [x] Foreman 3.4/Katello 4.6
* [x] Foreman 3.3/Katello 4.5
* [ ] Foreman 3.2/Katello 4.4
* [ ] Foreman 3.1/Katello 4.3
* For Foreman 3.0 or older, please create a separate PR.
* We do not accept PRs for Foreman 2.3 or older.

### CODE REVIEW

File: guides/common/modules/proc_mqtt-tuning.adoc [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+.*Procedure*
+. Enable the feature using: `satellite-installer --foreman-proxy-plugin-remote-execution-script-mode=pull-mqtt` (this will change the ReX transport mode for the Satellite or capsule from the default SSH one to MQTT). It’s important to mention that one Satellite/capsule server can only use one transport mode: SSH or MQTT
+
+. In order to bump the default number of clients, create the following system unit definition file and set its contents as specified in the below given example (this will allow the `mosquitto` server to handle ~5000 clients):
++
+[options="nowrap", subs="+quotes,verbatim,attributes"]
+----
+cat >/etc/systemd/system/mosquitto.service.d/limits.conf <<__EOF__
+[Service]
+LimitNOFILE=5000
```
Comment (ekohl) [inline guides/common/modules/proc_mqtt-tuning.adoc:19]: If this is common (and I'm quite sure it is), I'd prefer to support this properly in our installer. I'm worried our installer actually purges this file again.

Having said that: EL8 has a default file limit of 262144 so you're SERIOUSLY lowering it. See `systemctl show -p DefaultLimitNOFILE` and `systemctl show -p LimitNOFILE foreman-mosquitto.service` as well.

Are you sure you weren't testing this on EL7 which has a much lower limit of 4096?

Comment (Imaanpreet) [reply guides/common/modules/proc_mqtt-tuning.adoc:19]: Hello @ekohl , we are already calling it here [1] right? 

[1] https://github.com/theforeman/foreman-documentation/blob/master/guides/common/modules/con_performance-tuning-quickstart.adoc. 

@pablomh could you please answer about the 4096 limit thing ?

Comment (ekohl) [reply guides/common/modules/proc_mqtt-tuning.adoc:19]: What I mean is that in our documentation we want to avoid users managing individual files. If this is something that we want to support, the installer should have some option to do so. It may take tuning profiles into account, depending on how we choose to implement it.

File: guides/common/modules/proc_mqtt-tuning.adoc [hunk 2]
```diff
@@ -0,0 +1,28 @@
+[id="REX_pull_provider_{context}"]
```
Comment (maximiliankolb) [inline guides/common/modules/proc_mqtt-tuning.adoc:1]: Please ensure title, anchor, and file name match.

File: guides/common/modules/proc_mqtt-tuning.adoc [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+= REX pull provider 
+
+Remote Execution has been around for a while now and has served SSH-based and Ansible execution very well. In many environments these are not options due to either security or networking concerns.
+
+Katello-agent has provided an alternative for a long time that worked in an environment where SSH was not available or the Foreman/Katello server or smart proxies did not have the ability to connect to a client. However katello-agent only supported a small set of actions around updating packages. There has been a long desire to create a ‘pull’ based provider that has the features of REX but with a deployment model similar tokatello-agent. 
+
+The solution uses MQTT as its messaging protocol and includes an MQTT client running on each Host. In orderto tune the MQTT, you can follow the following procedure on the Satellite or capsule server:
+
+.*Procedure*
```
Comment (maximiliankolb) [inline guides/common/modules/proc_mqtt-tuning.adoc:10]:
```suggestion
.Procedure
```

It will automatically be bold.

File: guides/common/modules/proc_mqtt-tuning.adoc [hunk 4]
```diff
@@ -0,0 +1,28 @@
+[id="REX_pull_provider_{context}"]
+= REX pull provider 
+
+Remote Execution has been around for a while now and has served SSH-based and Ansible execution very well. In many environments these are not options due to either security or networking concerns.
```
Comment (maximiliankolb) [inline guides/common/modules/proc_mqtt-tuning.adoc:4]: Please stick to one sentence per line. For more information, see the CONTRIBUTING.md file.

**Review by maximiliankolb — CHANGES_REQUESTED** [review]

File: guides/common/modules/proc_mqtt-tuning.adoc (hunk 1 above)
Comment (pablomh) [reply guides/common/modules/proc_mqtt-tuning.adoc:19]: Hi @ekohl. In our `mosquitto` testing we found the following [issue] in a `medium` tuned (not that it matters in this case) capsule when trying to register the 1015th client:
~~~
Unable to accept new connection, system socket count has been exceeded. Try increasing "ulimit -n" or equivalent.
~~~
We were able to overcome the issue making the modifications that we propose in this PR, so maybe the `mosquitto` system unit is not inheriting properly that property.

[issue]: https://bugzilla.redhat.com/show_bug.cgi?id=2132774

Comment (ekohl) [reply guides/common/modules/proc_mqtt-tuning.adoc:19]: https://github.com/systemd/systemd/blob/23f3a6f5ff864fd26063c6c35fdaa6d85de566c7/NEWS#L5880-L5914 does have quite a bit to say on this. 

> Note that the soft limit remains at 1024 for compatibility reasons: the
          traditional UNIX select() call cannot deal with file descriptors >=
          1024 and increasing the soft limit globally might thus result in
          programs unexpectedly allocating a high file descriptor and thus
          failing abnormally when attempting to use it with select() (of
          course, programs shouldn't use select() anymore, and prefer
          poll()/epoll, but the call unfortunately remains undeservedly popular
          at this time).

Then later:

> Programs that want to take benefit of the increased limit have to "opt-in" into
          high file descriptors explicitly by raising their soft limit.

That 1015 is seriously close to 1024 that it makes me suspect it's observing the soft limit by default and I suspect that if Mosquitto was modified to raise its soft limit, it would automatically get a lot more file descriptors.

I think we should pursue the path of fixing it upstream first. For now I've opened https://github.com/eclipse/mosquitto/issues/2674 to see if it's a viable solution.

Comment (jhutar) [reply guides/common/modules/proc_mqtt-tuning.adoc:19]: Thank you Ewoud! Do we want to add a reference to that upstream ticket to the Tuning Guide in this PR?

File: guides/common/modules/proc_mqtt-tuning.adoc [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+The solution uses MQTT as its messaging protocol and includes an MQTT client running on each Host. In orderto tune the MQTT, you can follow the following procedure on the Satellite or capsule server:
+
+.*Procedure*
+. Enable the feature using: `satellite-installer --foreman-proxy-plugin-remote-execution-script-mode=pull-mqtt` (this will change the ReX transport mode for the Satellite or capsule from the default SSH one to MQTT). It’s important to mention that one Satellite/capsule server can only use one transport mode: SSH or MQTT
+
+. In order to bump the default number of clients, create the following system unit definition file and set its contents as specified in the below given example (this will allow the `mosquitto` server to handle ~5000 clients):
```
Comment (mjivraja) [inline guides/common/modules/proc_mqtt-tuning.adoc:13]:
```suggestion
. To bump the default number of clients, create the following system unit definition file and set its contents as specified in the below given example (this will allow the `mosquitto` server to handle ~5000 clients):
```

File: guides/common/modules/proc_mqtt-tuning.adoc [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+Remote Execution has been around for a while now and has served SSH-based and Ansible execution very well. In many environments these are not options due to either security or networking concerns.
+
+Katello-agent has provided an alternative for a long time that worked in an environment where SSH was not available or the Foreman/Katello server or smart proxies did not have the ability to connect to a client. However katello-agent only supported a small set of actions around updating packages. There has been a long desire to create a ‘pull’ based provider that has the features of REX but with a deployment model similar tokatello-agent. 
+
+The solution uses MQTT as its messaging protocol and includes an MQTT client running on each Host. In orderto tune the MQTT, you can follow the following procedure on the Satellite or capsule server:
+
+.*Procedure*
+. Enable the feature using: `satellite-installer --foreman-proxy-plugin-remote-execution-script-mode=pull-mqtt` (this will change the ReX transport mode for the Satellite or capsule from the default SSH one to MQTT). It’s important to mention that one Satellite/capsule server can only use one transport mode: SSH or MQTT
```
Comment (mjivraja) [inline guides/common/modules/proc_mqtt-tuning.adoc:11]:
```suggestion
. Enable the feature using: `satellite-installer --foreman-proxy-plugin-remote-execution-script-mode=pull-mqtt` (this will change the ReX transport mode for the Satellite or capsule from the default SSH one to MQTT). It’s important to mention that one Satellite or Capsule server can only use one transport mode: SSH or MQTT
```

File: guides/common/modules/proc_mqtt-tuning.adoc (hunk 2 above)
Comment (Imaanpreet) [reply guides/common/modules/proc_mqtt-tuning.adoc]: Made a change. Thank you.

File: guides/common/modules/proc_mqtt-tuning.adoc (hunk 3 above)
Comment (Imaanpreet) [reply guides/common/modules/proc_mqtt-tuning.adoc]: Made a change. Thank you.

File: guides/common/modules/proc_mqtt-tuning.adoc (hunk 4 above)
Comment (Imaanpreet) [reply guides/common/modules/proc_mqtt-tuning.adoc]: Made a change. Thank you.

File: guides/common/modules/proc_mqtt-tuning.adoc (hunk 5 above)
Comment (Imaanpreet) [reply guides/common/modules/proc_mqtt-tuning.adoc]: Made a change. Thank you.

File: guides/common/modules/proc_mqtt-tuning.adoc (hunk 6 above)
Comment (Imaanpreet) [reply guides/common/modules/proc_mqtt-tuning.adoc]: Made a change. Thank you.

**Comment by maximiliankolb** [comment]
@Imaanpreet Thanks for applying feedback. I've restarted the GHA pipeline. Please also have a look at used upstream terminology such as "Foreman" or "Katello" in your PR. In Foreman documentation, we generally use attributes, see https://github.com/theforeman/foreman-documentation/blob/master/guides/common/attributes-base.adoc

**Comment by jhutar** [comment]
Hello @maximiliankolb and @mjivraja . Please does this look OK now?

**Review by maximiliankolb — CHANGES_REQUESTED** [review]
This procedure still needs to make use of attributes, e.g. "satellite-installer" > "{foreman-installer}" and so on.

Also, it's not included, therefore we cannot see the built HTML output.

I also feel like the introduction is more of a background story then technical documentation. Would you mind rewording it?

Please also stick to "one sentence per line" as per our conventions.

**Comment by mjivraja** [comment]
> Hello @maximiliankolb and @mjivraja . Please does this look OK now?

Some comments are yet to be addressed. Can you have a look once again?

File: guides/common/modules/proc_rex-pull-based-provider.adoc [hunk 7]
```diff
@@ -0,0 +1,31 @@
+[id="REX_pull_based_provider_{context}"]
+= REX pull based provider
```
Comment (ehelms) [inline guides/common/modules/proc_rex-pull-based-provider.adoc:2]:
```suggestion
= Remote Execution pull-based transport
```

**Comment by Imaanpreet** [comment]
Hello @maximiliankolb & @mjivraja : could you please check now?

File: guides/common/modules/proc_rex-pull-based-provider.adoc [hunk 8]
```diff
@@ -0,0 +1,31 @@
+[id="REX_pull_based_provider_{context}"]
+= REX pull based provider 
+
+Remote Execution has been around for a while now and has served SSH-based execution very well.
```
[the remark this answers is not in the document]
Comment (ehelms) [reply guides/common/modules/proc_rex-pull-based-provider.adoc:4]: This is to define the acronym before we use it to make it clear what we are talking about.

File: guides/common/modules/proc_rex-pull-based-provider.adoc [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+ 
+Now, {Project} has a ‘pull’ based provider that has the features of REX but with a deployment model similar to katello-agent. 
+
+This provider uses MQTT as its messaging protocol and includes an MQTT client running on each content host.
+
+In order to tune the MQTT server (`mosquitto`) and increase the number of clients (content hosts) connnected to it, you can follow the following procedure on the {Project} server:
```
Comment (ehelms) [inline guides/common/modules/proc_rex-pull-based-provider.adoc:10]:
```suggestion
In order to tune the MQTT server (`mosquitto`) and increase the number of clients connected to it, you can follow the following procedure on the {Project} server:
```

File: guides/common/modules/proc_rex-pull-based-provider.adoc [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+
+.Procedure
+. Enable the feature using: `{foreman-installer} --foreman-proxy-plugin-remote-execution-script-mode pull-mqtt` (this will change the REX transport mode from the default SSH one to MQTT).
+  It's important to mention that one {Project} server ({ProjectServer} or {SmartProxyServer}) can only use one transport mode: SSH or MQTT.
+
+. To increase the default number of clients accepted by the MQTT server, create the following system unit definition file and set its contents as specified in the below given example (this will allow the `mosquitto` service to handle ~5000 clients):
```
Comment (ehelms) [inline guides/common/modules/proc_rex-pull-based-provider.adoc:16]:
```suggestion
. To increase the default number of hosts accepted by the MQTT server, create the following system unit definition file and set its contents as specified in the below given example (this will allow the `mosquitto` service to handle ~5000 clients):
```

File: guides/common/modules/proc_rex-pull-based-provider.adoc [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+[Service]
+LimitNOFILE=5000
+__EOF__
+----
+. Once the file has been edited, the following commands need to be run to make the tuning come into effect:
++
+[options="nowrap", subs="+quotes,verbatim,attributes"]
+----
+systemctl daemon-reload
+systemctl restart mosquitto.service
```
Comment (maximiliankolb) [inline guides/common/modules/proc_rex-pull-based-provider.adoc:30]:
```suggestion
# systemctl daemon-reload
# systemctl restart mosquitto.service
```

File: guides/common/modules/proc_rex-pull-based-provider.adoc [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+. To increase the default number of clients accepted by the MQTT server, create the following system unit definition file and set its contents as specified in the below given example (this will allow the `mosquitto` service to handle ~5000 clients):
++
+[options="nowrap", subs="+quotes,verbatim,attributes"]
+----
+cat >/etc/systemd/system/mosquitto.service.d/limits.conf <<__EOF__
+[Service]
+LimitNOFILE=5000
+__EOF__
+----
+. Once the file has been edited, the following commands need to be run to make the tuning come into effect:
```
Comment (maximiliankolb) [inline guides/common/modules/proc_rex-pull-based-provider.adoc:25]:
```suggestion
. Run the following commands apply your changes:
```

File: guides/common/modules/proc_rex-pull-based-provider.adoc [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+Now, {Project} has a ‘pull’ based provider that has the features of REX but with a deployment model similar to katello-agent. 
+
+This provider uses MQTT as its messaging protocol and includes an MQTT client running on each content host.
+
+In order to tune the MQTT server (`mosquitto`) and increase the number of clients (content hosts) connnected to it, you can follow the following procedure on the {Project} server:
+
+.Procedure
+. Enable the feature using: `{foreman-installer} --foreman-proxy-plugin-remote-execution-script-mode pull-mqtt` (this will change the REX transport mode from the default SSH one to MQTT).
+  It's important to mention that one {Project} server ({ProjectServer} or {SmartProxyServer}) can only use one transport mode: SSH or MQTT.
+
```
Comment (maximiliankolb) [inline guides/common/modules/proc_rex-pull-based-provider.adoc:15]:
```suggestion
```

File: guides/common/modules/proc_rex-pull-based-provider.adoc [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+ 
+Now, {Project} has a ‘pull’ based provider that has the features of REX but with a deployment model similar to katello-agent. 
+
+This provider uses MQTT as its messaging protocol and includes an MQTT client running on each content host.
+
+In order to tune the MQTT server (`mosquitto`) and increase the number of clients (content hosts) connnected to it, you can follow the following procedure on the {Project} server:
+
+.Procedure
+. Enable the feature using: `{foreman-installer} --foreman-proxy-plugin-remote-execution-script-mode pull-mqtt` (this will change the REX transport mode from the default SSH one to MQTT).
+  It's important to mention that one {Project} server ({ProjectServer} or {SmartProxyServer}) can only use one transport mode: SSH or MQTT.
```
Comment (maximiliankolb) [inline guides/common/modules/proc_rex-pull-based-provider.adoc:14]:
```suggestion
. Enable pull-based remote execution on your {ProjectServer} or {SmartProxyServer}:
+
[options="nowrap", subs="+quotes,verbatim,attributes"]
----
{foreman-installer} --foreman-proxy-plugin-remote-execution-script-mode pull-mqtt
----
+
Note that your {ProjectServer} or {SmartProxyServer} can only use one transport mode, either SSH or MQTT.
```

Maybe reword slightly here and there; but in general, we use code blocks instead of inline commands.

File: guides/common/modules/proc_rex-pull-based-provider.adoc (hunk 10 above)
Comment (maximiliankolb) [reply guides/common/modules/proc_rex-pull-based-provider.adoc:16]: Is this optional? What's the default value? Is 5000 clients a recommendation? Does it depends on the specs/used tuning profile?

**Review by maximiliankolb — CHANGES_REQUESTED** [review]
Made some suggestions.

File: guides/common/modules/proc_rex-pull-based-provider.adoc (hunk 12 above)
Comment (pablomh) [reply guides/common/modules/proc_rex-pull-based-provider.adoc:25]: s/apply/to apply/

File: guides/common/modules/proc_rex-pull-based-provider.adoc (hunk 10 above)
Comment (adamruzicka) [reply guides/common/modules/proc_rex-pull-based-provider.adoc:16]: It is optional, but there might be dragons, depending on your scale. According to systemd docs[1] the default limit is 1024. Based on the reasoning in[1], I wouldn't say it depends on specs/tuning profile.

[1] - https://www.freedesktop.org/software/systemd/man/systemd.exec.html#Process%20Properties

Comment (ekohl) [reply guides/common/modules/proc_rex-pull-based-provider.adoc:16]: Please see https://github.com/theforeman/foreman-documentation/pull/1747#discussion_r1009725326 for more discussion. I'd say that if we raise beyond 5000, we might as well raise it WAY higher. Then if we do want to limit it, we should use `max_connections` in `mosquitto.conf` instead.

Comment (pablomh) [reply guides/common/modules/proc_rex-pull-based-provider.adoc:16]: We chose 5000 because it's what we were able to test, but agree with your reasoning.

File: guides/common/modules/proc_rex-pull-based-provider.adoc (hunk 12 above)
Comment (Imaanpreet) [reply guides/common/modules/proc_rex-pull-based-provider.adoc]: done. thank you.

**Comment by Imaanpreet** [comment]
Hello @maximiliankolb : I have applied some more changes as per the suggestions. Could you please check now? Please correct me if I have missed anything. Thank you in advance :)

**Review by maximiliankolb — CHANGES_REQUESTED** [review]
A few more comments:

Right now, the first four lines are four paragraphs. It's probably best to remove the empty lines 5, 7, and 9 so that it'll be one paragraph.
Second, you have to actually include your procedure in an assembly. Right now, your content is not part of any guide.
Also, please ensure the file name, anchor, and title match.

**Comment by Imaanpreet** [comment]
> A few more comments:
> 
> Right now, the first four lines are four paragraphs. It's probably best to remove the empty lines 5, 7, and 9 so that it'll be one paragraph. Second, you have to actually include your procedure in an assembly. Right now, your content is not part of any guide. Also, please ensure the file name, anchor, and title match.

I hope now made changes will help.

**Comment by maximiliankolb** [comment]
Looks like there is a merge conflict right now. Please rebase to master & then we can trigger the GHA so you'll see your change rendered.

**Comment by Imaanpreet** [comment]
> Looks like there is a merge conflict right now. Please rebase to master & then we can trigger the GHA so you'll see your change rendered.

done. could you please check now.

File: guides/common/modules/proc_rex-pull-based-transport.adoc [hunk 15]
```diff
@@ -0,0 +1,34 @@
+[id="REX_pull_based_transport_{context}"]
+= REX pull-based transport
+
+Remote Execution (REX) has been around for a while now and has served SSH-based execution very well.
```
Comment (ehelms) [inline guides/common/modules/proc_rex-pull-based-transport.adoc:4]: I think we can drop this line, as it won't age well release to release.

File: guides/common/modules/proc_rex-pull-based-transport.adoc [hunk 16]
```diff
@@ -0,0 +1,34 @@
+[id="REX_pull_based_transport_{context}"]
+= REX pull-based transport
+
+Remote Execution (REX) has been around for a while now and has served SSH-based execution very well. 
+Now, {Project} has a pull-based transport that has the features of REX but with a deployment model similar to katello-agent.
```
Comment (ehelms) [inline guides/common/modules/proc_rex-pull-based-transport.adoc:5]: I think we should drop the temporal part of this statement:

```suggestion
{Project} has a pull-based transport that has the features of REX but with a deployment model similar to katello-agent. 
```

**Comment by ehelms** [comment]
@Imaanpreet If you could address my two comments, and then also squash into a single commit we should be good to go.

**Comment by Imaanpreet** [comment]
> @Imaanpreet If you could address my two comments, and then also squash into a single commit we should be good to go.

done @ehelms 
thank you

**Comment by Imaanpreet** [comment]
@ehelms : Hello, could you please check once? I have all the commits in one commit. 
https://github.com/theforeman/foreman-documentation/pull/1747/commits/77b2446dad7edac262ae81270458c18a7db4f586

**Review by ehelms — APPROVED** [review]

**Comment by ehelms** [comment]
This can be merged -- I am not doing so at present as I cannot open the cherry picks presently.

**Comment by maximiliankolb** [comment]
Thanks @Imaanpreet for your PR! Merged to master and cherry-picked: :heavy_check_mark: 
   84a5bd094..9fe3e5fa8  3.5 -> 3.5
   a351fd5da..60b7f60da  3.4 -> 3.4
   0f6281418..ac07b9bee  3.3 -> 3.3
